File: (03-23-2024)3Sum/joshlin/solution.ts
function threeSum(nums: number[]): number[][] {
    nums = nums.sort((a, b) => a - b);
    const res: number[][] = [];
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] > 0) break;
        if (i > 0 && nums[i - 1] == nums[i]) continue;
        let l = i + 1;
        let r = nums.length - 1;
        while (l < r) {
            let target = nums[i] + nums[l] + nums[r];
            if (target > 0) {
                r -= 1;
            } else if (target < 0) {
                l += 1;
            } else {
                res.push([nums[i], nums[l], nums[r]]);
                while (l < r && nums[l + 1] == nums[l]) l += 1;
                while (l < r && nums[r - 1] == nums[r]) r -= 1;
                l += 1;
                r -= 1;
            };
            
        };
    };
    return res;
};

File: (03-23-2024)3Sum/joshlin/solution.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        res = []
        for i in range(len(nums) - 2):
            if nums[i] > 0:
                break
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            l, r = i + 1, len(nums) - 1
            while l < r:
                target = nums[i] + nums[l] + nums[r]
                if target > 0:
                    r -= 1
                elif target < 0:
                    l += 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l + 1] == nums[l]:
                        l += 1
                    while l < r and nums[r - 1] == nums[r]:
                        r -= 1
                    l += 1
                    r -= 1
        return res